packages/imitation_game/README.md

```diff
@@ -0,0 +1,89 @@
+# Imitation Game
+
+## Description
+
+`imitation_game` is a platform for performing automated tests to compare Flutter
+to alternative platforms.  For example, how much memory does the same app
+written in Flutter and UIKit take?
+
+## Running all the tests
+
+You need a mobile device plugged into your computer and setup for development.
+The mobile device and the computer need to be on the same network, one that
+allows communication between computers since that's how the mobile phone will
+report its results to the computer.
+
+```sh
+dart imitation_game.dart
+```
+
+## Dependencies
+
+In order to run the tests you will need the union of all the platforms being
+tested.  As new tests are added please add to this list:
+
+### iOS
+
+- Flutter
+- Xcode
+- [ios_deploy](https://github.com/ios-control/ios-deploy) - used to launch apps
+  on the attached iOS device.
+
+## File Layout
+
+```
+./
+├─ imitation_game.dart
+└─ tests/
+   ├─ smiley/
+   │  ├─ README.md
+   │  ├─ flutter/
+   │  │  └─ run_ios.sh
+   │  └─ uikit/
+   │     └─ run_ios.sh
+   └─ memory/
+      ├─ README.md
+      ├─ flutter/
+      │  └─ run_ios.sh
+      └─ uikit/
+         └─ run_ios.sh
+```
+
+Here there are 2 different tests with 2 different platform implementations.  The
+tests are named `smiley` and `memory`, they are both implemented on the
+platforms `flutter` and `uikit`.
+
+### Adding a test
+
+Tests should comprise of implementations on one or more platform.  The directory
+for the test should be added to `./tests`.  Inside that directory there should
+be a directory of implementations and a `README.md` file that explains the test.
+
+### Adding an implementation to a test
+
+An implementation has to follow these rules:
+
+- It needs to perform the same operations as the other implementations and
+  follow the description in the test's `README.md`.
+- It needs to contain a `run_ios.sh` script that will build and launch the test
+  on the connected device.
+- It should contain a file named `ip.txt` which will be overwritten by
```

Comment:
I still feel that it's better to put the device-to-host reporting logic inside the test (e.g., `smiley`) so different target OSes and tests could choose reporting mechanisms differently. For example, for Android memory reporting, it's probably easier to just use `adb` on the host side so no HTTP communication is required from the device side. If multiple tests do want to use the HTTP reporting, then they could all import the same HTTP reporting module?

Reply:
There is value in having the reporting mechanism the same across all implementations.  HTTP is the most universal mechanism that should be easy for users.  Once we have a few tests we can probably just wrap it up in a library so when you implement a test you just call `reportResult(result);` and you don't even care how it's getting reported.

Reply:
I like just calling `reportResult` without having to worry about HTTP. It feels better to me if we don't put this HTTP communication as a contract or API of the imitation game so we'll have the freedom to change.